File: franchise_erp/franchise_erp/report/custom_counter_store_level_performance/custom_counter_store_level_performance.js
// Copyright (c) 2026, Franchise Erp and contributors
// For license information, please see license.txt

frappe.query_reports["Custom Counter Store-Level Performance"] = {

	filters: [

		{
			fieldname: "sales_manager",
			label: __("Sales Manager"),
			fieldtype: "Link",
			options: "User"
		},

		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			reqd: 1,
			default: frappe.datetime.month_start()
		},

		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			reqd: 1,
			default: frappe.datetime.month_end()
		}

	],


	// =====================================================
	// FORMATTER
	// =====================================================

	formatter: function (
		value,
		row,
		column,
		data,
		default_formatter
	) {

		value = default_formatter(
			value,
			row,
			column,
			data
		);


		// =================================================
		// SALES MANAGER
		// =================================================

		if (
			column.fieldname === "sales_manager" &&
			data &&
			data.sales_manager
		) {

			return `
				<a
					href="#"
					class="sales-manager-link"
					data-user="${encodeURIComponent(
						data.sales_manager
					)}"
					style="
						font-weight: 600;
						cursor: pointer;
					"
				>
					${frappe.utils.escape_html(
						data.sales_manager
					)}
				</a>
			`;
		}


		// =================================================
		// COUNTER COUNT
		// =================================================

		if (
			column.fieldname === "counter_count" &&
			data
		) {

			return `
				<a
					href="#"
					class="counter-count-link"
					data-counter-details="${encodeURIComponent(
						data.counter_details || "[]"
					)}"
					style="
						cursor: pointer;
						font-weight: 600;
					"
				>
					${data.counter_count || 0}
				</a>
			`;
		}


		return value;
	},


	// =====================================================
	// ONLOAD
	// =====================================================

	onload: function () {


		// =================================================
		// SALES MANAGER CLICK
		// =================================================

		$(document)
			.off(
				"click.custom_counter_report_manager",
				".sales-manager-link"
			)
			.on(
				"click.custom_counter_report_manager",
				".sales-manager-link",
				function (e) {

					e.preventDefault();
					e.stopPropagation();

					const user =
						decodeURIComponent(
							$(this).attr(
								"data-user"
							) || ""
						);

					open_sales_manager_delivery_notes(
						user
					);
				}
			);


		// =================================================
		// COUNTER COUNT CLICK
		// =================================================

		$(document)
			.off(
				"click.custom_counter_report",
				".counter-count-link"
			)
			.on(
				"click.custom_counter_report",
				".counter-count-link",
				function (e) {

					e.preventDefault();
					e.stopPropagation();

					let counters = [];

					try {

						const counter_details =
							decodeURIComponent(
								$(this).attr(
									"data-counter-details"
								) || "%5B%5D"
							);

						counters =
							JSON.parse(
								counter_details
							);

					} catch (error) {

						console.error(
							"Unable to parse counter details:",
							error
						);

						frappe.msgprint(
							__(
								"Unable to load counter details."
							)
						);

						return;
					}


					if (
						!counters.length
					) {

						frappe.msgprint(
							__(
								"No counters found."
							)
						);

						return;
					}


					show_counter_dialog(
						counters
					);
				}
			);
	}
};


// =========================================================
// OPEN SALES MANAGER DELIVERY NOTES
// =========================================================
function open_sales_manager_delivery_notes(user) {

	const from_date =
		frappe.query_report.get_filter_value("from_date");

	const to_date =
		frappe.query_report.get_filter_value("to_date");

	if (!user || !from_date || !to_date) {
		frappe.msgprint(
			__("User, From Date and To Date are required.")
		);
		return;
	}

	// =====================================================
	// GET CURRENT SALES MANAGER ROW
	// =====================================================

	const report_data = frappe.query_report.data || [];

	const manager_row = report_data.find(
		row => row.sales_manager === user
	);

	if (!manager_row) {
		frappe.msgprint(
			__("Sales Manager data not found.")
		);
		return;
	}

	// =====================================================
	// GET ASSIGNED COUNTERS
	// =====================================================

	let counters = [];

	try {
		counters = JSON.parse(
			manager_row.counter_details || "[]"
		);
	} catch (error) {
		console.error(
			"Unable to parse counter details:",
			error
		);

		frappe.msgprint(
			__("Unable to load assigned counters.")
		);

		return;
	}

	// =====================================================
	// GET UNIQUE COMPANIES
	// =====================================================

	const companies = [
		...new Set(
			counters
				.map(counter =>
					(counter.represents_company || "").trim()
				)
				.filter(Boolean)
		)
	];

	if (!companies.length) {
		frappe.msgprint(
			__("No company found for assigned counters.")
		);
		return;
	}

	// =====================================================
	// OPEN DELIVERY NOTE REPORT
	// =====================================================

	const params = new URLSearchParams();

	/*
	 * IMPORTANT:
	 * Frappe MultiSelect/List filter format:
	 *
	 * company = ["in", ["Company 1", "Company 2"]]
	 */

	params.set(
		"company",
		JSON.stringify([
			"in",
			companies
		])
	);

	params.set(
		"docstatus",
		"1"
	);

	params.set(
		"posting_date",
		JSON.stringify([
			"Between",
			[
				from_date,
				to_date
			]
		])
	);

	const url =
		"/app/delivery-note/view/report?" +
		params.toString();

	console.log("Companies:", companies);
	console.log("Delivery Note URL:", url);

	window.open(
		url,
		"_blank"
	);
}


// =========================================================
// SHOW COUNTER DIALOG
// =========================================================

function show_counter_dialog(
	counters
) {

	const dialog =
		new frappe.ui.Dialog({

			title: __(
				"Assigned Counters"
			),

			size: "small",

			fields: [
				{
					fieldname:
						"counter_list",

					fieldtype:
						"HTML"
				}
			]
		});


	let html = "";


	if (
		!counters ||
		!counters.length
	) {

		html = `
			<div class="text-muted">
				${__(
					"No counters found."
				)}
			</div>
		`;

	} else {

		html = `
			<div class="counter-list">

				${counters
					.map(
						(
							counter,
							index
						) => {

							const customer_name =
								frappe.utils.escape_html(
									counter.customer_name ||
									counter.name ||
									""
								);


							const customer_id =
								frappe.utils.escape_html(
									counter.name ||
									""
								);


							const company =
								frappe.utils.escape_html(
									counter.represents_company ||
									""
								);


							const counter_owner =
								frappe.utils.escape_html(
									counter.counter_owner ||
									""
								);


							return `
								<div
									style="
										padding: 12px 4px;
										border-bottom:
											${
												index ===
												counters.length - 1
													? "none"
													: "1px solid var(--border-color)"
											};
									"
								>

									<a
										href="#"
										class="counter-dn-link"
										data-customer="${encodeURIComponent(
											counter.name || ""
										)}"
										style="
											font-weight: 600;
											cursor: pointer;
										"
									>
										${customer_name}
									</a>


									${
										customer_id &&
										customer_id !==
											customer_name
											? `
												<div
													class="text-muted"
													style="
														font-size: 12px;
														margin-top: 2px;
													"
												>
													${customer_id}
												</div>
											`
											: ""
									}


									${
										company
											? `
												<div
													class="text-muted"
													style="
														font-size: 12px;
														margin-top: 2px;
													"
												>
													${__(
														"Company"
													)}:
													${company}
												</div>
											`
											: ""
									}


									${
										counter_owner
											? `
												<div
													class="text-muted"
													style="
														font-size: 12px;
														margin-top: 2px;
													"
												>
													${__(
														"Created By"
													)}:
													${counter_owner}
												</div>
											`
											: ""
									}

								</div>
							`;
						}
					)
					.join("")}

			</div>
		`;
	}


	dialog.fields_dict
		.counter_list
		.$wrapper
		.html(
			html
		);


	// =====================================================
	// COUNTER CLICK
	// =====================================================

	dialog.fields_dict
		.counter_list
		.$wrapper

		.off(
			"click",
			".counter-dn-link"
		)

		.on(
			"click",
			".counter-dn-link",
			function (e) {

				e.preventDefault();

				const customer =
					decodeURIComponent(
						$(this).attr(
							"data-customer"
						) || ""
					);


				if (!customer) {

					frappe.msgprint(
						__(
							"Customer is not set for this Counter."
						)
					);

					return;
				}


				open_counter_delivery_notes(
					customer
				);
			}
		);


	dialog.show();
}


// =========================================================
// OPEN COUNTER DELIVERY NOTES
// =========================================================

function open_counter_delivery_notes(
	customer
) {

	const from_date =
		frappe.query_report.get_filter_value(
			"from_date"
		);

	const to_date =
		frappe.query_report.get_filter_value(
			"to_date"
		);


	if (!customer) {

		frappe.msgprint(
			__(
				"Customer is not set for this Counter."
			)
		);

		return;
	}


	if (
		!from_date ||
		!to_date
	) {

		frappe.msgprint(
			__(
				"Please select From Date and To Date."
			)
		);

		return;
	}


	const params =
		new URLSearchParams();


	// =================================================
	// Specific Counter / Customer
	// =================================================

	params.set(
		"customer",
		customer
	);


	params.set(
		"docstatus",
		"1"
	);


	params.set(
		"posting_date",
		JSON.stringify([
			"Between",
			[
				from_date,
				to_date
			]
		])
	);


	const url =
		"/app/delivery-note/view/report?" +
		params.toString();


	window.open(
		url,
		"_blank"
	);
}

File: franchise_erp/franchise_erp/report/custom_counter_store_level_performance/custom_counter_store_level_performance.py
# Copyright (c) 2026, Franchise Erp and contributors
# For license information, please see license.txt

import re

import frappe
from frappe import _
from frappe.utils import flt


def execute(filters=None):
    filters = frappe._dict(filters or {})

    columns = get_columns()
    data = get_data(filters)

    return columns, data


# =========================================================
# COLUMNS
# =========================================================

def get_columns():
    return [
        {
            "label": _("Sales Manager"),
            "fieldname": "sales_manager",
            "fieldtype": "Link",
            "options": "User",
            "width": 200,
        },
        {
            "label": _("Sales Manager Name"),
            "fieldname": "sales_manager_name",
            "fieldtype": "Data",
            "width": 180,
        },
        {
            "label": _("No. of Counters"),
            "fieldname": "counter_count",
            "fieldtype": "Int",
            "width": 130,
        },
        {
            "label": _("Counter Details"),
            "fieldname": "counter_details",
            "fieldtype": "Data",
            "hidden": 1,
        },
        {
            "label": _("Total Sales"),
            "fieldname": "total_sales",
            "fieldtype": "Currency",
            "width": 160,
        },
        {
            "label": _("Incentive %"),
            "fieldname": "incentive_percentage",
            "fieldtype": "Percent",
            "width": 120,
        },
        {
            "label": _("Total Incentive Earned"),
            "fieldname": "total_incentive_earned",
            "fieldtype": "Currency",
            "width": 190,
        },
    ]


# =========================================================
# MAIN DATA
# =========================================================
def get_data(filters):
    if not filters.get("from_date") or not filters.get("to_date"):
        return []

    incentive_slabs = get_counter_incentive_slabs()

    conditions = [
        "IFNULL(TRIM(c.account_manager), '') != ''"
    ]

    values = {}

    # =========================================================
    # SALES MANAGER FILTER
    # =========================================================

    if filters.get("sales_manager"):
        conditions.append(
            "TRIM(c.account_manager) = %(sales_manager)s"
        )

        values["sales_manager"] = (
            filters.sales_manager.strip()
        )

    # =========================================================
    # FETCH COUNTERS
    # =========================================================

    counters = frappe.db.sql(
        f"""
        SELECT
            c.name,
            c.customer_name,
            TRIM(c.account_manager) AS account_manager,
            c.represents_company
        FROM `tabCustomer` c
        WHERE
            {" AND ".join(conditions)}
        ORDER BY
            c.account_manager,
            c.customer_name
        """,
        values,
        as_dict=True,
    )

    # =========================================================
    # GROUP COUNTERS BY SALES MANAGER
    # =========================================================

    manager_groups = {}

    for counter in counters:

        manager = (
            counter.account_manager or ""
        ).strip()

        if not manager:
            continue

        if manager not in manager_groups:
            manager_groups[manager] = []

        manager_groups[manager].append(
            {
                "name": counter.name,
                "customer_name": counter.customer_name,
                "represents_company": (
                    counter.represents_company
                ),
            }
        )

    data = []

    # =========================================================
    # ONE ROW PER SALES MANAGER
    # =========================================================

    for manager, manager_counters in manager_groups.items():

        counter_count = len(manager_counters)

        # Assigned counters ki represents_company values
        companies = list(
            dict.fromkeys(
                counter["represents_company"].strip()
                for counter in manager_counters
                if counter.get("represents_company")
                and counter["represents_company"].strip()
            )
        )

        # Delivery Note company ke basis par Total Sales
        total_sales = get_total_sales(
            companies,
            filters.from_date,
            filters.to_date,
        )

        incentive = get_incentive_percentage(
            total_sales,
            incentive_slabs,
        )

        payout = calculate_payout(
            total_sales,
            incentive,
        )

        sales_manager_name = (
            frappe.db.get_value(
                "User",
                manager,
                "full_name",
            )
            or ""
        )

        data.append(
            {
                "sales_manager": manager,
                "sales_manager_name": sales_manager_name,
                "counter_count": counter_count,
                "counter_details": frappe.as_json(
                    manager_counters
                ),
                "total_sales": total_sales,
                "incentive_percentage": incentive,
                "total_incentive_earned": payout,
            }
        )
    return data

# =========================================================
# GET SALES BY OWNER
# =========================================================

def get_sales_by_owner(
    owner,
    from_date,
    to_date,
):
    """
    Calculate total submitted Delivery Note sales
    for a specific user/owner.
    """

    if not owner:
        return 0

    result = frappe.db.sql(
        """
        SELECT
            COALESCE(
                SUM(dn.net_total),
                0
            ) AS total_sales

        FROM `tabDelivery Note` dn

        WHERE
            dn.owner = %(owner)s
            AND dn.docstatus = 1
            AND dn.posting_date >= %(from_date)s
            AND dn.posting_date <= %(to_date)s
        """,
        {
            "owner": owner,
            "from_date": from_date,
            "to_date": to_date,
        },
        as_dict=True,
    )

    if not result:
        return 0

    return flt(
        result[0].total_sales
    )


# =========================================================
# GET TOTAL SALES
# =========================================================
# =========================================================
# GET TOTAL SALES
# =========================================================
def get_total_sales(
    companies,
    from_date,
    to_date,
):
    """
    Calculate Total Sales from Delivery Notes.

    Logic:

    Sales Manager
        ↓
    Assigned Counter
        ↓
    Counter.represents_company
        ↓
    Delivery Note.company
        ↓
    SUM(Delivery Note.net_total)
    """

    if not companies:
        return 0

    if isinstance(companies, str):
        companies = [companies]

    companies = list(
        dict.fromkeys(
            company.strip()
            for company in companies
            if company and company.strip()
        )
    )

    if not companies:
        return 0

    result = frappe.db.sql(
        """
        SELECT
            COALESCE(
                SUM(dn.net_total),
                0
            ) AS total_sales

        FROM `tabDelivery Note` dn

        WHERE
            dn.docstatus = 1

            AND dn.company IN %(companies)s

            AND dn.posting_date BETWEEN
                %(from_date)s
                AND %(to_date)s
        """,
        {
            "companies": tuple(companies),
            "from_date": from_date,
            "to_date": to_date,
        },
        as_dict=True,
    )

    if not result:
        return 0

    return flt(
        result[0].total_sales
    )

# =========================================================
# CALCULATE PAYOUT
# =========================================================

def calculate_payout(
    total_sales,
    incentive,
):
    """
    Calculate incentive payout.

    Example:

        Sales = 40,546
        Incentive = 0.1%

        Payout = 40.55
    """

    if incentive is None:
        return 0

    return (
        flt(total_sales)
        * flt(incentive)
        / 100
    )


# =========================================================
# GET INCENTIVE SLABS
# =========================================================

def get_counter_incentive_slabs():

    child_table_doctype = frappe.db.get_value(
        "DocField",
        {
            "parent": "TZU Setting",
            "fieldname": "counter_store_level_performance",
            "fieldtype": "Table",
        },
        "options",
    )

    if not child_table_doctype:
        return []

    rows = frappe.get_all(
        child_table_doctype,

        filters={
            "parent": "TZU Setting",
            "parenttype": "TZU Setting",
            "parentfield": "counter_store_level_performance",
        },

        fields=[
            "idx",
            "aggregate_sales",
            "incentive",
        ],

        order_by="idx asc",
    )

    slabs = []

    for row in rows:

        aggregate_sales = parse_amount(
            row.aggregate_sales
        )

        if aggregate_sales is None:
            continue

        slabs.append(
            {
                "idx": row.idx,
                "aggregate_sales": aggregate_sales,
                "incentive": flt(
                    row.incentive
                ),
            }
        )

    # -----------------------------------------------------
    # Sort by aggregate sales
    # -----------------------------------------------------

    slabs.sort(
        key=lambda slab: (
            slab["aggregate_sales"],
            slab["idx"],
        )
    )

    return slabs


# =========================================================
# PARSE AMOUNT
# =========================================================

def parse_amount(value):

    if value is None:
        return None

    value = str(value).strip()

    if not value:
        return None

    # Remove currency symbol
    value = value.replace(
        "₹",
        "",
    )

    # Remove commas
    value = value.replace(
        ",",
        "",
    )

    # Remove spaces
    value = re.sub(
        r"\s+",
        "",
        value,
    )

    # Extract numbers
    numbers = re.findall(
        r"\d+(?:\.\d+)?",
        value,
    )

    if not numbers:
        return None

    return flt(
        numbers[0]
    )


# =========================================================
# GET INCENTIVE PERCENTAGE
# =========================================================

def get_incentive_percentage(
    total_sales,
    incentive_slabs,
):
    """
    Find the highest applicable incentive slab.

    Example:

        0 - 100000      -> 0%
        100000 - 200000 -> 0.5%
        200000+         -> 1%

    If sales = 250000
    incentive = 1%
    """

    if not incentive_slabs:
        return None

    matched_incentive = None

    for slab in incentive_slabs:

        aggregate_target = (
            slab["aggregate_sales"]
        )

        if total_sales >= aggregate_target:

            matched_incentive = (
                slab["incentive"]
            )

        else:
            break

    return matched_incentive